### PR TITLE
systemd: use sd-switch

### PR DESCRIPTION
**Pull request by rycee**
This makes the systemd module use the sd-switch application to perform the unit switch during a generation activation.

Since the closure of sd-switch is relatively light-weight we unconditionally pull it in as a dependency. We simultaneously remove the `systemd.user.startServices` option and perform the switch action automatically.

### CODE REVIEW

File: modules/systemd.nix [hunk 1]
```diff
@@ -99,6 +99,11 @@ let
       ) + "\n";
     };
 
+  sd-switch =
+    let
+      url = "https://gitlab.com/rycee/sd-switch/-/archive/master/sd-switch-master.tar.bz2";
```
Comment (cole-h) [inline modules/systemd.nix:104]: Will this be pinned at some point (e.g. when this PR is no longer a draft/considered "ready")? Rubs me the wrong way to depend on a moving target like this.

Comment (rycee) [reply modules/systemd.nix:104]: Yes, this PR can only be merged once this program has been in Nixpkgs for some time.

Comment (cole-h) [reply modules/systemd.nix:104]: Great, thanks for the clarification. I have at least one MR I'll be making to `sd-switch` in the coming days (or when I can find time) -- I love the choice to write it in Rust ;^)

**Comment by rycee** [comment]
Rebased to master in 9c0fe3957bb91128d5d3599e289fe9344a293ece.

**Comment by misuzu** [comment]
`sd-switch` is causing issues on headless machines (nixpkgs: https://github.com/NixOS/nixpkgs/commit/840c782d507d60aaa49aa9e3f6d0b0e780912742, home-manager: https://github.com/rycee/home-manager/commit/9c0fe3957bb91128d5d3599e289fe9344a293ece):
```
> RUST_BACKTRACE=full VERBOSE=1 DRY_RUN=1 /nix/store/98gv4dj2zarcxqpv1gyxd2982yavz7rj-home-manager-generation/activate
Starting home manager activation
Sanity checking Nix
Sanity checking oldGenNum and oldGenPath
This is a dry run
Using Nix version: nix-env (Nix) 2.3.7
Activation variables:
  oldGenNum=108
  oldGenPath=/nix/store/98gv4dj2zarcxqpv1gyxd2982yavz7rj-home-manager-generation
  newGenPath=/nix/store/98gv4dj2zarcxqpv1gyxd2982yavz7rj-home-manager-generation
  newGenNum=109
  newGenGcPath=/nix/var/nix/gcroots/per-user/misuzu/current-home
  genProfilePath=/nix/var/nix/profiles/per-user/misuzu/home-manager
Activating checkFilesChanged
Activating checkLinkTargets
Activating writeBoundary
Activating installPackages
Activating linkGeneration
Cleaning up orphan links from /home/misuzu
Checking /home/misuzu/.piknik.toml: exists
Checking /home/misuzu/.config/mc/ini: exists
Checking /home/misuzu/.config/fontconfig/conf.d/10-hm-fonts.conf: exists
Checking /home/misuzu/.config/broot/conf.toml: exists
Checking /home/misuzu/.config/broot/launcher/installed-v1: exists
Checking /home/misuzu/.config/git/ignore: exists
Checking /home/misuzu/.config/git/config: exists
Checking /home/misuzu/.config/htop/htoprc: exists
Checking /home/misuzu/.config/starship.toml: exists
Checking /home/misuzu/.zshrc: exists
Checking /home/misuzu/.cache/.keep: exists
Checking /home/misuzu/.dir_colors: exists
Checking /home/misuzu/.tmux.conf: exists
No change so reusing latest profile generation 108
Creating home file links in /home/misuzu
mkdir -p --verbose /home/misuzu
ln -nsf --verbose /nix/store/67xy9s6lkanc52wh5d3ij2zx76sn51ja-home-manager-files/.piknik.toml /home/misuzu/.piknik.toml
mkdir -p --verbose /home/misuzu/.config/mc
ln -nsf --verbose /nix/store/67xy9s6lkanc52wh5d3ij2zx76sn51ja-home-manager-files/.config/mc/ini /home/misuzu/.config/mc/ini
mkdir -p --verbose /home/misuzu/.config/fontconfig/conf.d
ln -nsf --verbose /nix/store/67xy9s6lkanc52wh5d3ij2zx76sn51ja-home-manager-files/.config/fontconfig/conf.d/10-hm-fonts.conf /home/misuzu/.config/fontconfig/conf.d/10-hm-fonts.conf
mkdir -p --verbose /home/misuzu/.config/broot
ln -nsf --verbose /nix/store/67xy9s6lkanc52wh5d3ij2zx76sn51ja-home-manager-files/.config/broot/conf.toml /home/misuzu/.config/broot/conf.toml
mkdir -p --verbose /home/misuzu/.config/broot/launcher
ln -nsf --verbose /nix/store/67xy9s6lkanc52wh5d3ij2zx76sn51ja-home-manager-files/.config/broot/launcher/installed-v1 /home/misuzu/.config/broot/launcher/installed-v1
mkdir -p --verbose /home/misuzu/.config/git
ln -nsf --verbose /nix/store/67xy9s6lkanc52wh5d3ij2zx76sn51ja-home-manager-files/.config/git/ignore /home/misuzu/.config/git/ignore
mkdir -p --verbose /home/misuzu/.config/git
ln -nsf --verbose /nix/store/67xy9s6lkanc52wh5d3ij2zx76sn51ja-home-manager-files/.config/git/config /home/misuzu/.config/git/config
mkdir -p --verbose /home/misuzu/.config/htop
ln -nsf --verbose /nix/store/67xy9s6lkanc52wh5d3ij2zx76sn51ja-home-manager-files/.config/htop/htoprc /home/misuzu/.config/htop/htoprc
mkdir -p --verbose /home/misuzu/.config
ln -nsf --verbose /nix/store/67xy9s6lkanc52wh5d3ij2zx76sn51ja-home-manager-files/.config/starship.toml /home/misuzu/.config/starship.toml
mkdir -p --verbose /home/misuzu
ln -nsf --verbose /nix/store/67xy9s6lkanc52wh5d3ij2zx76sn51ja-home-manager-files/.zshrc /home/misuzu/.zshrc
mkdir -p --verbose /home/misuzu/.cache
ln -nsf --verbose /nix/store/67xy9s6lkanc52wh5d3ij2zx76sn51ja-home-manager-files/.cache/.keep /home/misuzu/.cache/.keep
mkdir -p --verbose /home/misuzu
ln -nsf --verbose /nix/store/67xy9s6lkanc52wh5d3ij2zx76sn51ja-home-manager-files/.dir_colors /home/misuzu/.dir_colors
mkdir -p --verbose /home/misuzu
ln -nsf --verbose /nix/store/67xy9s6lkanc52wh5d3ij2zx76sn51ja-home-manager-files/.tmux.conf /home/misuzu/.tmux.conf
Activating onFilesChange
Activating reloadSystemD
Options are Options {
    dry_run: true,
    verbose: true,
    timeout: None,
    old_dir: Some(
        "/nix/store/98gv4dj2zarcxqpv1gyxd2982yavz7rj-home-manager-generation/home-files/.config/systemd/user",
    ),
    new_dir: "/nix/store/98gv4dj2zarcxqpv1gyxd2982yavz7rj-home-manager-generation/home-files/.config/systemd/user",
}
Performing dry-run
Enabling verbose output
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: DBus(D-Bus error: Unable to autolaunch a dbus-daemon without a $DISPLAY for X11 (org.freedesktop.DBus.Error.NotSupported))', src/lib.rs:349:27
stack backtrace:
   0:     0x55e3284fe6a2 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hbc432fcb96ef6975
   1:     0x55e328522fbc - core::fmt::write::h8410abd894bca434
   2:     0x55e32850b077 - std::io::Write::write_fmt::h86de6a3429010a4b
   3:     0x55e328506455 - std::panicking::default_hook::{{closure}}::h50ddb77147ffcd59
   4:     0x55e32850613f - std::panicking::default_hook::hf23e22c4e3e68213
   5:     0x55e328506a48 - std::panicking::rust_panic_with_hook::h9bdeb7f9ba3511ec
   6:     0x55e32850664b - rust_begin_unwind
   7:     0x55e328522ca1 - core::panicking::panic_fmt::h8b4a8da40d3433a9
   8:     0x55e328522d33 - core::result::unwrap_failed::h91aa5ce6338c5a8d
   9:     0x55e3284d15e3 - sd_switch::switch::h3795b09053fd25f6
  10:     0x55e3284c8506 - sd_switch::main::h0eb390f483c5a989
  11:     0x55e3284c8f33 - std::rt::lang_start::{{closure}}::h18ca04172d5eebb6
  12:     0x55e328506d35 - std::rt::lang_start_internal::h057182eff4078461
  13:     0x55e3284c8f22 - main
  14:     0x7f438f7e0c7d - __libc_start_main
  15:     0x55e3284c71da - _start
  16:                0x0 - <unknown>
```

**Comment by rycee** [comment]
@misuzu Ah sorry for the inconvenience, I really should have realized this issue with dbus. I think it is pretty fatal so I've reverted the commit until the dbus setup in NixOS has been fixed or sd-switch has been rewritten not to use dbus. Thanks for the quick and detailed report!

**Pull request closed without merging**